### PR TITLE
refactor: Separate ui rendering from data updates

### PR DESCRIPTION
**Pull request by almet**
`U.SCHEMA` now contains an `impacts` key, which makes it possible to specify what part of the UI is impacted by data changes.

A new `render` method has been added on `U.Map` and `U.DataLayer`, which is used to rerender the proper parts of the UI depending on the passed properties.

`U.FormBuilder` calls this `render()` method (if present), during form changes.

> [!Note]
> This approach is less specific than [a previous one](https://github.com/umap-project/umap/compare/master...feat/simple-crdts) in order to a) not tie the schema to specific callbacks and b) not repeat properties in two different places.
> 
> The resulting code is less conservative when considering which parts of the UI are updated when receiving updates, thus rerendering parts of the UI which might not be in need of rerender. The hope is that it won't be a problem, but let's keep our eyes open.

## Progress

- [x] Sort schema keys
- [x] Add all UI impacts on schema keys
- [x] Update FormBuilder instanciations to remove callbacks when possible
- [x] Ensure the tests pass
- [x] `ChoroplethLayer.onEdit` is expecting to be passed the builder
- [x] Use `impacts = []` everywhere (remove `['']` occurrences)
- [x] Change how the functional tests are working (see #1718 #1719) 
- [x] Make the JS unit tests pass
- [x] Make the functional tests pass

### CODE REVIEW

File: umap/static/umap/js/modules/schema.js [hunk 1]
```diff
@@ -1,45 +1,52 @@
...
+  datalayersControl: {
+    impacts: ['ui'],
+    type: Boolean,
+    nullable: true,
+    handler: 'DataLayersControl',
+    label: translate('Display the data layers control'),
+    default: true,
   },
   defaultView: {
+    impacts: [], // no need to update the ui, only useful when loading the map
```
Comment (davidbgk) [inline umap/static/umap/js/modules/schema.js:49]: Is there a difference between `['']` (above) and `[]`? Maybe update the `// Possible impacts` comment to explain that?

File: umap/static/umap/js/umap.js [hunk 2]
```diff
@@ -240,6 +241,42 @@ U.Map = L.Map.extend({
     this.on('click contextmenu.show', this.closeInplaceToolbar)
   },
 
+  render: function (fields) {
+    let impacts = U.Utils.getImpactsFromSchema(fields)
```
Comment (davidbgk) [inline umap/static/umap/js/umap.js:245]: If the schema is static, don't we want to compute that once and for all? (vs. at each `render()` call)

File: umap/static/umap/js/modules/schema.js (hunk 1 above)
Comment (almet) [reply umap/static/umap/js/modules/schema.js:49]: No, there shouldn't be, and the list should be updated to only have `[]`.

File: umap/static/umap/js/umap.js (hunk 2 above)
Comment (almet) [reply umap/static/umap/js/umap.js:245]: The schema is static but the list of passed fields isn't.`getImpactsFromSchema` goes trough the keys of the schema, returning what is impacted by the given fields.

I don't think the lookup will take much time. If it is, we might want to cache out the results, maybe. I'd say it's an optimisation for the future, when/if we need it. Would that be okay for you?

Comment (davidbgk) [reply umap/static/umap/js/umap.js:245]: Alright 👍

File: umap/static/umap/js/modules/utils.js [hunk 3]
```diff
@@ -22,3 +22,20 @@ export function checkId(string) {
   if (typeof string !== 'string') return false
   return /^[A-Za-z0-9]{5}$/.test(string)
 }
+
+export function getImpactsFromSchema(fields) {
```
Comment (yohanboniface) [inline umap/static/umap/js/modules/utils.js:26]: I'd make a few unitests (JS) for this one :)

File: umap/static/umap/js/umap.js [hunk 4]
```diff
@@ -240,6 +241,42 @@ U.Map = L.Map.extend({
...
+          break
+      }
+    }
+  },
+
+  reindexEachDataLayer: function () {
+    this.eachDataLayer((datalayer) => datalayer.reindex())
+  },
+
+  renderVisibleDataLayers: function () {
```
Comment (yohanboniface) [inline umap/static/umap/js/umap.js:274]: `redrawVisibleDataLayers` ?

File: umap/static/umap/js/umap.layer.js [hunk 5]
```diff
@@ -594,6 +596,31 @@ U.DataLayer = L.Evented.extend({
     if (this.autoLoaded()) this.map.on('zoomend', this.onZoomEnd, this)
   },
 
+  render: function (fields, builder) {
```
Comment (yohanboniface) [inline umap/static/umap/js/umap.layer.js:599]: Maybe we could call the method `onDataChanged` to make it more explicit that this is not the default redering but a sort of "reflow" ?

File: umap/static/umap/js/modules/schema.js [hunk 6]
```diff
@@ -1,45 +1,52 @@
...
-  miniMap: {
-    type: Boolean,
-    label: translate('Do you want to display a minimap?'),
-    default: false,
-  },
-  displayPopupFooter: {
-    type: Boolean,
-    label: translate('Do you want to display popup footer?'),
-    default: false,
+  color: {
```
Comment (yohanboniface) [inline umap/static/umap/js/modules/schema.js:24]: I suggest to add a dedicated `test_on_edit.py` file with some basic test (at least one per "impact"), that we can then complete later with some edge cases.

File: umap/static/umap/js/umap.forms.js [hunk 7]
```diff
@@ -1033,6 +1033,7 @@ U.FormBuilder = L.FormBuilder.extend({
   setter: function (field, value) {
     L.FormBuilder.prototype.setter.call(this, field, value)
     if (this.options.makeDirty !== false) this.obj.isDirty = true
+    if ('render' in this.obj) this.obj.render([field], this)
```
Comment (yohanboniface) [inline umap/static/umap/js/umap.forms.js:1036]: this may call render on map when searching in the browser. So either this is only done when `makeDirty` is false, either we use a dedicated Builder class for the browser (simpler imho).

File: umap/static/umap/js/umap.js [hunk 8]
```diff
@@ -240,6 +241,42 @@ U.Map = L.Map.extend({
...
+          this.initTileLayers()
+          break
+        case 'bounds':
+          this.handleLimitBounds()
+          break
+      }
+    }
+  },
+
+  reindexEachDataLayer: function () {
```
Comment (yohanboniface) [inline umap/static/umap/js/umap.js:270]: `reindexDataLayers` ?

File: umap/static/umap/js/umap.layer.js [hunk 9]
```diff
@@ -254,16 +254,18 @@ U.Layer.Choropleth = L.FeatureGroup.extend({
   },
 
   onEdit: function (field, builder) {
+    // Only compute the breaks if we're dealing with choropleth
+    if (!field.startsWith('options.choropleth')) return
     // If user touches the breaks, then force manual mode
     if (field === 'options.choropleth.breaks') {
       this.datalayer.options.choropleth.mode = 'manual'
-      builder.helpers['options.choropleth.mode'].fetch()
+      if (builder) builder.helpers['options.choropleth.mode'].fetch()
```
Comment (yohanboniface) [inline umap/static/umap/js/umap.layer.js:262]: Ugly but let's go with that for now! ;)

**Review by yohanboniface — COMMENTED** [review]
:+1:

File: umap/static/umap/js/umap.forms.js (hunk 7 above)
Comment (almet) [reply umap/static/umap/js/umap.forms.js:1036]: I don't think we should rely on `makeDirty` as I think it will be out of the picture at some point.

Using a dedicated browser class seem the way to go :-)

File: umap/static/umap/js/umap.js (hunk 8 above)
Comment (almet) [reply umap/static/umap/js/umap.js:270]: 👍

File: umap/static/umap/js/umap.js (hunk 4 above)
Comment (almet) [reply umap/static/umap/js/umap.js:274]: 👍

File: umap/static/umap/js/umap.layer.js (hunk 5 above)
Comment (almet) [reply umap/static/umap/js/umap.layer.js:599]: I would prefer to keep the `on*` methods for events. (`onDataChanged` would imply there is an `data-changed` event that occurred). 

I like the idea of using `render` somewhere, because it's really the intent here. Maybe `rerenderFields`? (it also  makes it more explicit that's it's expecting fields to be passed).

File: umap/static/umap/js/umap.forms.js (hunk 7 above)
Comment (almet) [reply umap/static/umap/js/umap.forms.js:1036]: In the end, I did the opposite: this call only happens if `makeDirty` is not explicitely set to `false`.

From where I stand right now, it seems to avoid adding complexity for no almost no benefits. We can take care of doing things differently if/when we remove the `dirty` mechanism.

File: umap/static/umap/js/modules/utils.js (hunk 3 above)
Comment (almet) [reply umap/static/umap/js/modules/utils.js]: ✔️

**Comment by almet** [comment]
This is ready to be merged!

File: umap/static/umap/js/modules/schema.js [hunk 10]
```diff
@@ -1,46 +1,61 @@
...
   },
-  scaleControl: {
+  captionBar: {
     type: Boolean,
-    label: translate('Do you want to display the scale control?'),
-    default: true,
+    impacts: ['ui'],
+    label: translate('Do you want to display a caption bar?'),
+    default: false,
+    impacts: [],
```
Comment (davidbgk) [inline umap/static/umap/js/modules/schema.js:16]: Two different `impacts` keys for the captionBar?

Comment (almet) [reply umap/static/umap/js/modules/schema.js]: 🤫

File: umap/tests/integration/test_on_edit.py [hunk 11]
```diff
@@ -0,0 +1,185 @@
+import re
```
Comment (yohanboniface) [inline umap/tests/integration/test_on_edit.py:1]: Maybe mv those tests in `test_edit_map.py` instead of having a new module ?

File: umap/static/umap/test/Util.js [hunk 12]
```diff
@@ -1,397 +1,4 @@
 describe('L.Util', function () {
```
Comment (yohanboniface) [inline umap/static/umap/test/Util.js:1]: I'm surprised to see all those tests removed here ? Should this have been done in that other PR moving utils to a module ? Or is this just missing a rebase ?

File: umap/static/umap/js/umap.layer.js (hunk 5 above)
Comment (yohanboniface) [reply umap/static/umap/js/umap.layer.js:599]: ok, let's keep `render`!

**Review by yohanboniface — APPROVED** [review]

File: umap/static/umap/test/Util.js (hunk 12 above)
Comment (almet) [reply umap/static/umap/test/Util.js:1]: yep, it's broken in the current master, and fixed here! 🙈
